### PR TITLE
[2nd] Load textures from the subfolders in texturepacks

### PR DESCRIPTION
**Pull request by numberZero**
Rebase&update of #4468 
*Not tested!*

EDIT by paramat: Addresses #2821 #4464

### CODE REVIEW

File: src/server.cpp [hunk 1]
```diff
@@ -251,8 +251,14 @@ Server::Server(
 
 	// Apply texture overrides from texturepack/override.txt
 	std::string texture_path = g_settings->get("texture_path");
-	if (!texture_path.empty() && fs::IsDir(texture_path))
-		m_nodedef->applyTextureOverrides(texture_path + DIR_DELIM + "override.txt");
+	if (!texture_path.empty() && fs::IsDir(texture_path)) {
+		std::vector<std::string> paths;
+		paths.push_back(texture_path);
+		std::set<char> chars_to_ignore = { '_', '.' };
+		fs::GetRecursiveSubPaths(texture_path, paths, false, &chars_to_ignore);;
```
Comment (SmallJoker) [inline src/server.cpp]: 1) spotted `;;`
2) Could you please move the block, start=`std::string texture_path`, end=`fs::GetRecursiveSubPaths` into a function? An empty vector can be returned if the path is not set or not a directory.

**Comment by numberZero** [comment]
@SmallJoker

File: src/filesys.cpp [hunk 2]
```diff
@@ -380,15 +380,33 @@ std::string TempPath()
...
+}
+
+std::vector<std::string> GetRecursiveDirs(const std::string& dir)
+{
+	std::vector<std::string> result;
+	GetRecursiveDirs(result, dir);
+	return result;
+}
+
+void GetRecursiveSubPaths(const std::string &path, std::vector<std::string> &dst, bool list_files, const std::set<char> *ignore)
```
Comment (SmallJoker) [inline src/filesys.cpp]: Line too long, > 80 characters

File: src/filesys.h [hunk 3]
```diff
@@ -66,10 +67,20 @@ bool DeleteSingleFileOrEmptyDirectory(const std::string &path);
...
 /* Multiplatform */
 
-// The path itself not included
-void GetRecursiveSubPaths(const std::string &path, std::vector<std::string> &dst);
+/* The path itself not included, returns a list of all subpaths.
+   dst - vector that contains all the subpaths.
+   list files - include files in the list of subpaths.
+   ignore - paths that start with these charcters will not be listed.
+*/
+void GetRecursiveSubPaths(const std::string &path, std::vector<std::string> &dst, bool list_files, const std::set<char> *ignore);
```
Comment (SmallJoker) [inline src/filesys.h]: Line too long, > 80 characters

File: src/filesys.cpp [hunk 4]
```diff
@@ -380,15 +380,33 @@ std::string TempPath()
 
 #endif
 
-void GetRecursiveSubPaths(const std::string &path, std::vector<std::string> &dst)
+void GetRecursiveDirs(std::vector<std::string>& dirs, const std::string& dir)
```
Comment (SmallJoker) [inline src/filesys.cpp]: `std::vector<std::string> &dirs, const std::string &dir`

File: src/filesys.cpp [hunk 5]
```diff
@@ -380,15 +380,33 @@ std::string TempPath()
...
+{
+	static const std::set<char> chars_to_ignore = { '_', '.' };
+	if (dir.empty() || !IsDir(dir))
+		return;
+	dirs.push_back(dir);
+	fs::GetRecursiveSubPaths(dir, dirs, false, &chars_to_ignore);
+}
+
+std::vector<std::string> GetRecursiveDirs(const std::string& dir)
```
Comment (SmallJoker) [inline src/filesys.cpp]: `const std::string &dir`

Comment (numberZero) [reply src/filesys.cpp]: er, my IDE did that...

**Comment by numberZero** [comment]
fixed

File: src/client/tile.cpp [hunk 6]
```diff
@@ -129,11 +129,12 @@ std::string getTexturePath(const std::string &filename)
...
 		Check from texture_path
 	*/
-	const std::string &texture_path = g_settings->get("texture_path");
-	if (!texture_path.empty()) {
-		std::string testpath = texture_path + DIR_DELIM + filename;
+	for (const auto &path : fs::GetRecursiveDirs(g_settings->get("texture_path"))) {
+		std::string testpath = path + DIR_DELIM + filename;
 		// Check all filename extensions. Returns "" if not found.
 		fullpath = getImagePath(testpath);
+		if (fullpath != "")
```
Comment (nerzhul) [inline src/client/tile.cpp]: !fullpath.empty

File: src/client/tile.cpp [hunk 7]
```diff
@@ -129,11 +129,12 @@ std::string getTexturePath(const std::string &filename)
 	/*
 		Check from texture_path
 	*/
-	const std::string &texture_path = g_settings->get("texture_path");
-	if (!texture_path.empty()) {
-		std::string testpath = texture_path + DIR_DELIM + filename;
+	for (const auto &path : fs::GetRecursiveDirs(g_settings->get("texture_path"))) {
```
Comment (nerzhul) [inline src/client/tile.cpp]: is this possible to cache fs::GetRecursiveDirs for this value ? it's called many times

Comment (numberZero) [reply src/client/tile.cpp]: Possible. Where could be the variable defined? It is used both in the server and in the client.

File: src/client/tile.cpp (hunk 6 above)
Comment (numberZero) [reply src/client/tile.cpp]: ok

File: src/client/tile.cpp (hunk 7 above)
Comment (numberZero) [reply src/client/tile.cpp]: @nerzhul

**Comment by SmallJoker** [comment]
@nerzhul please refresh your review.

File: src/filesys.cpp [hunk 8]
```diff
@@ -380,15 +380,36 @@ std::string TempPath()
...
+{
+	std::vector<std::string> result;
+	GetRecursiveDirs(result, dir);
+	return result;
+}
+
+void GetRecursiveSubPaths(const std::string &path,
+		  std::vector<std::string> &dst,
+		  bool list_files,
+		  const std::set<char> *ignore)
```
Comment (sfan5) [inline src/filesys.cpp]: couldn't this be a `const std::set<char> &ignore`?

Comment (numberZero) [reply src/filesys.cpp]: Done. @sfan5

**Comment by paramat** [comment]
05d93c7fa1be9245dd5211b7dc1bdf0961b39eea

**Comment by Fixer-007** [comment]
There was report on IRC recently, did not verified it:
http://irc.minetest.net/minetest-dev/2017-11-16#i_5138309
>23:37 	Wayward_One 	after git pulling, none of my texture packs are showing up..
>23:43 	Wayward_One 	i can only select them now by manually specifying a certain texture's path in minetest.conf

**Comment by paramat** [comment]
@numberZero

**Comment by numberZero** [comment]
That is, TP works, but not shown in the list? I don’t see any changes in the code that might have such effect. I hope I will be able to look into that soon.

**Comment by Fixer-007** [comment]
List of packs shows up but when I select some - default textures are still been used.

**Comment by paramat** [comment]
False alarm perhaps? http://irc.minetest.net/minetest-dev/2017-11-18#i_5139116

**Comment by numberZero** [comment]
> looks like a settings problem, as closing & re-opening Minetest takes the new chosen TP
> "static thread_local std::vector<std::string> dirs"
> perhaps a little too much of caching?

Yet another setting requiring restart. #5462

**Comment by Fixer-007** [comment]
Not false alarm, it seems to activate a texture pack you need full restart of the entire application. That's clearly a bug, should I open one?

**Comment by tacotexmex** [comment]
Is this going into the 0.4 backport?

**Comment by nerzhul** [comment]
no. 0.4 is bugfix only

**Pull request closed without merging**